langchain4j-core/src/main/java/dev/langchain4j/web/search/WebSearchResults.java

```diff
@@ -0,0 +1,166 @@
+package dev.langchain4j.web.search;
+
+import dev.langchain4j.data.document.Document;
+import dev.langchain4j.data.segment.TextSegment;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static dev.langchain4j.internal.ValidationUtils.ensureNotEmpty;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Represents the response of a web search performed.
+ * This includes the list of organic search results, information about the search, and pagination information.
+ * <p>
+ * {@link WebSearchResults} follow opensearch foundation standard implemented by most web search engine libs like Google, Bing, Yahoo, etc.
+ * <a href="https://github.com/dewitt/opensearch/blob/master/opensearch-1-1-draft-6.md#examples-of-opensearch-responses">OpenSearch#response</a>
+ * </p>
+ * <p>
+ * The organic search results are the web pages that are returned by the search engine in response to a search query.
+ * These results are typically ranked by relevance to the search query.
+ */
+public class WebSearchResults {
+
+    private final Map<String, Object> searchMetadata;
```

Comment:
What is the difference between `searchMetadata` here and `searchInformation.metadata`?

Reply:
Sorry if it is not very clear in the documentation. `searchMetadata` is additional information about the execution context of the web search at the time of performing the search. `searchInformation.metadata` is every information associated with the search query, each web search implementation could set more values that `totalResults` and `pageNumber` (fields by default) there if they think needed.

E.g.
`searchMetadata`,  SerpAPI returns the status (success, failed, etc), the final url formed and the execution time, Google API returns something similar, another web search can return other fields that is why the `Map<String, Object>`.
`searchInformation.metadata` returns the total results of the query, total number of pages of the query. Other searcher implementations may have other relevant values that could store it in that (if there images, promotions, etc).

Reply:
Perharps I still don't fully grasp it... But I still can't understand why can't we merge `searchMetadata`, `searchInformation.metadata` and `searchInformation.xyz` fields together in a single map for simplicity? To me it sounds the same "metadata"...
Could you please explain?

Reply:
both are metadata at different levels. e.g.:
`search_metadata` = Search engine execution info. More related to each search engine.
`search_information.metadata` = Info related to the search user query, which aren't part of a web page result.
```
{
  "search_metadata": { // info about search execution
    "status": "Success",
    "processed_at": "2022-07-18 23:42:02 UTC",
    "total_time_taken": 1.35s
  },
  "search_information": {  // info about user query results
     "total_results": 343004
     "page_number": 1
     "metadata": {
        "contain_images": true,
        "images": [
         "url 1",
         "...."
        ] 
     }
   }
```
We can simplify it into a single metadata, yes, but I would leave it to various web search implementations to be able to map the data they need. SerpAPI for example response lots of data...